bootloader/piboot.go

```diff
@@ -0,0 +1,441 @@
+	if output, err := exec.Command("sh", "-c",
+		"mv "+bcomFiles+" "+dstDir).CombinedOutput(); err != nil {
+		return fmt.Errorf("cannot move RPi dtbs to %s:\n%s",
+			dstDir, output)
+	}
+	overlaysDir := filepath.Join(dstDir, "dtbs/overlays/")
+	newOvDir := filepath.Join(dstDir, "overlays/")
+	if err := os.Rename(overlaysDir, newOvDir); err != nil {
+		if !os.IsExist(err) {
+			return err
+		}
+	}
+
+	// README file is needed so os_prefix is honored for overlays. See
+	// https://www.raspberrypi.com/documentation/computers/config_txt.html#os_prefix
+	readmeOverlays, err := os.Create(filepath.Join(dstDir, "overlays", "README"))
+	if err != nil {
+		return err
+	}
+	readmeOverlays.Close()
+	return nil
+}
+
+func (p *piboot) ExtractKernelAssets(s snap.PlaceInfo, snapf snap.Container) error {
+	// Rootdir will point to ubuntu-boot, but we need to put things in ubuntu-seed
+	dstDir := filepath.Join(ubuntuSeedDir, pibootPartFolder, s.Filename())
+
+	logger.Debugf("ExtractKernelAssets to %s", dstDir)
+
+	return p.layoutKernelAssetsToDir(snapf, dstDir)
+}
+
+func (p *piboot) ExtractRecoveryKernelAssets(recoverySystemDir string, s snap.PlaceInfo,
+	snapf snap.Container) error {
+	if recoverySystemDir == "" {
+		return fmt.Errorf("internal error: recoverySystemDir unset")
+	}
+
+	recoveryKernelAssetsDir :=
+		filepath.Join(p.rootdir, recoverySystemDir, "kernel")
+	logger.Debugf("ExtractRecoveryKernelAssets to %s", recoveryKernelAssetsDir)
+
+	if err := p.layoutKernelAssetsToDir(snapf, recoveryKernelAssetsDir); err != nil {
+		return err
+	}
+	return p.loadAndApplyConfig()
+}
+
+func (p *piboot) RemoveKernelAssets(s snap.PlaceInfo) error {
+	return removeKernelAssetsFromBootDir(
+		filepath.Join(ubuntuSeedDir, pibootPartFolder), s)
+}
+
+// Finds out the kernel loaded from a given RPi config file
+func (p *piboot) getKernelFromRpiConfig(rpiCfgFn string) (snap.PlaceInfo, error) {
+	rpiCfgPth := filepath.Join(ubuntuSeedDir, rpiCfgFn)
+	buf, err := ioutil.ReadFile(rpiCfgPth)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, ErrNoTryKernelRef
+		}
+		return nil, err
+	}
+
+	lines := strings.Split(string(buf), "\n")
+
+	prefix := ""
+	for _, line := range lines {
+		if strings.HasPrefix(line, "os_prefix=") {
+			keyVal := strings.Split(line, "=")
+			prefix = keyVal[1]
+			break
+		}
+	}
+	if prefix == "" {
+		return nil, fmt.Errorf("no os_prefix in %s", rpiCfgFn)
+	}
+
+	kernelSnapFileName := filepath.Base(prefix)
+	sn, err := snap.ParsePlaceInfoFromSnapFileName(kernelSnapFileName)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"cannot parse kernel snap file name from prefix %q: %v",
+			prefix,
+			err,
+		)
+	}
+	logger.Debugf("kernel from %s is %s", rpiCfgFn, kernelSnapFileName)
+	return sn, nil
+}
+
+func (p *piboot) EnableKernel(sinfo snap.PlaceInfo) error {
+	p.SetBootVars(map[string]string{"snap_kernel": sinfo.Filename()})
+	return p.loadAndApplyConfig()
+}
+
+func (p *piboot) EnableTryKernel(sinfo snap.PlaceInfo) error {
```

Comment:
So after reading this implementation and thinking about it more, I think we should scrap my idea about trying to implement ExtractedRunKernelImageBootloader, since we really don't actually have an extracted run kernel image here for one, and for two it's just adding unnecessary complexity and confusion here I think. The main problem with implementing that method is that the methods from that interface are called in different orders with respect to SetBootVars() for different setups and I think that leaks too much complexity into this method, when we could just as well just skip implementing those methods and contain all the complexity inside SetBootVars itself.

So this is what I am thinking now for implementing SetBootVars (note my other comment about always using just one bootenv file on ubuntu-seed, so we don't have multiple bootenv files):
* For all vars, first set the key/value in the piboot.cfg file we use as the bootenv, but that the pi bootloader itself actually just totally ignores
* If we are setting snapd_recovery_mode==run, then also configure config.txt and cmdline.txt to boot from the extracted kernel referenced by snap_kernel
* If we are settting snapd_recovery_mode==install|run, then also configure config.txt and cmdline.txt to boot from the extracted seed system at /systems/${snapd_recovery_system}/... 
* If setting kernel_status=try, then copy config.txt to tryboot.txt and set the os_prefix to use the new kernel referenced by snap_try_kernel
* If setting kernel_status="", if tryboot.txt exists, and the new value of snap_kernel matches the information in tryboot.txt, then replace config.txt with tryboot.txt (note that if we go with my kernel command line parameter suggestion for implementing getting kernel_status in GetBootVars, then here we would need to remove that tryboot-cmdline.txt setting `snap_tried_kernel=1` too). If tryboot.txt does exist or the new value of snap_kernel is not the kernel referenced in tryboot.txt (i.e. this is the case in a failed boot where we set kernel_status="" and snap_try_kernel="" but don't modify snap_kernel at all), then just delete tryboot.txt. If tryboot.txt doesn't exist then this does nothing extra AFAICT

For implementing GetBootVars, see my other suggestion about the two possibilites (/run file populated in the initramfs or a kernel command line specifically for the try kernel